### ISSUE TITLE
use of pseudo-labels in tfrec creation

### ISSUE THREAD

**Issue by tfriedel** [HelenaELiu/FROST#3]
I've been going through the code an in the TF version.
In the create_split() method in frost.py the records with the highest confidence scores are written to a new '-label.tfrecord' file. I would assume that the labels in that file should be pseudo-labels. 
But it seems like only the *indices* of those images are used to find the records to copy from the original '-train.tfrecord'. 
``` py 
        with tf.python_io.TFRecordWriter(target + '-label.tfrecord') as writer_label:
            pos, loop = 0, trange(count, desc='Writing records')
            #for infile in input_file:
            for record in tf.python_io.tf_record_iterator(input_file):
                if pos in label:
                    writer_label.write(record)
```
Shouldn't the the records be modified to contain pseudo-labels instead of real labels? I fear that this might be a bug and it might be responsible for the very good results we see. 
But maybe I'm missing something ?

**Comment by lnsmith54** [comment]
Thank you for your attentiveness to our code.  It will help us to get it right.

I believe the code is correct as it is.  The method mimics the script that is located at scripts/create_splits.py.  As you can see from the data setup instructions, this script creates the *labeled* datasets.  Similarly, the method in frost.py creates the labeled datasets.  

The difference between the method and the script is where the labels come from.  This code works by organizing the training samples by class order.  If you look a few lines earlier in the method, you will see:
``` 
       # Distribute labels to match the input distribution.
        train_data = np.load(target + ".npy")
        label = []
        for i in range(size):
            c = i%nclass
            npos = i//nclass
            label.append(train_data[c][npos])
```
This loads the sample number and pseudo-labels that were saved by the bootstrap method.  The labels used here are the pseudo-labels because the training data was sorted into class by pseudo-labels in bootstrap.

**Comment by tfriedel** [comment]
Yes, the training data was sorted into class by pseudo-labels. So the row in the train_data indicates the pseudo-label that should be used. But what is written to the tfrecord is the unchanged record from input_file, which contains no pseudo-labels, but real labels. A record contains the fields "image" and "label". We can assume that a pseudo-label not always matches the real-label, so there needs to be some overwriting of the actual label with pseudolabels before writing the records. 
I.e something like
```example.features.feature['label'].int64_list.value[0] = pseudolabel```
I looked into the tfrecords that are written, for example a file like "./HakaVusN/cifar10.1@80-label.tfrecord". ALL entries there have original labels in the "label" field. So it's no wonder that in the end the accuracies are within 1% of the accuracies you get with supervised learning, because actual labels and not pseudo labels are being used. 
Another way you could test if this is happening is by not providing any labels beyond the initial 1-label per class.

**Comment by lnsmith54** [comment]
Yesterday I looked into this matter and discovered that you are correct.  That was a good catch and my apologies for my error (it was due to my ignorance of TFrecords).  I worked on a fix and thoroughly tested it.  I am going to rerun all of the FROST experiments and will update the paper appropriately to the new results.  We also plan to update the code in this repository later today.

**Comment by Piotrm777** [comment]
May I ask , where I can find new version of paper with updated results ? 
Thanks in advance

**Comment by lnsmith54** [comment]
We have withdrawn the paper because of the impact the error has on the results.  We are making new investigations but these will take time.  Thank you for your patience.

**Comment by oguiza** [comment]
Thanks for your research in this area and your honesty. It’s greatly appreciated.
You just mention Tensorflow in your response. Could you please confirm if the Pytorch code is also impacted by this issue?

**Comment by lnsmith54** [comment]
Thank you for your comments.  We appreciate them.

Yes, the PyTorch code had the same error.  We've corrected the mistakes and the results are significantly lower.  We are considering a couple of alternative methods in SSL that seem promising and we are hopeful.
